### PR TITLE
Fix "buildcache update-index --keys ..." when mirror is S3

### PR DESCRIPTION
**Pull request by scottwittenburg**
Calling `generate_key_index` with no temporary directory argument results in failure and a stack trace when the mirror is on S3.  This change attempts to handle the default `tmpdir` parameter of `None` a little better.

### CODE REVIEW

**Comment by scottwittenburg** [comment]
@opadron Now that I'm in that code a little more, I feel there's some weirdness around the use of a temporary directory.  I see how you only wanted to create a single temporary directory and use it both in `push_keys` and `generate_key_index`, but that makes it awkward to call `generate_key_index` from anywhere else.  I don't think every call site should have to replicate the logic of checking whether the mirror is local or remote and pass a temporary directory if it's remote, especially since that check is done inside the `generate_key_index` method anyway.

As I'm thinking about it more, it may be the problem is really that anytime we want to `push_to_url` we need to pass it the path to a local file.  If that method could take a file-like object or even a string, we could avoid the need to always open a temp file, write to it, then push that.  

Any thoughts?

**Review by opadron — APPROVED** [review]
Approved.

**Comment by opadron** [comment]
@scottwittenburg Regarding your question about accepting opened file-like objects:

Yes, the code could be changed so that acceptance of an opened file-like object could be provided on a best-effort basis.  The trade off is that it would complicate the web and url abstractions in Spack, because what "best-effort" means is going to depend on the URL scheme.

For example, `s3` uploads using `boto3` require that such objects are opened in binary mode ([ref](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-uploading-files.html)).  So for `s3`, "best-effort" means using the native upload when we can, and either making another file-like object that wraps the given one or just falling back to the current make-a-temporary-copy-and-upload-that strategy.  What's more, we can almost certainly count on new URL schemes being introduced at some point (there have already been questions about adding `ssh://` and `gcp://` schemes, for example), and so handling this issue generically would require extensions to the current abstraction that would allow implementations to communicate what tools are available to the generic caller and how that caller should consider them in determining what "best-effort" means for the current call.

For now, the caller is (non-generically) avoiding tmpdirs for local filesystem mirrors (since the file can be written directly), and writing a temporary copy to use as the source for everything else (assuming that every other scheme will represent some kind of non-local resource; something that can't be directly written to).  For those URLs, we avoid tmpdirs if its caller passes one along, and only accept filenames rather than opened file-like objects, since it is the lowest common denominator in what we expect implementations to accept as a source.

EDIT: At least, that's how it will work once this is merged!  :)

**Comment by scottwittenburg** [comment]
@becker33 Getting coverage up for this PR would, I think, involve some mockery of S3 mirrors which I don't think we have yet.  How do you feel about me merging this in spite of that low coverage?  The change is very small and would result in the `develop` pipelines passing again on gitlab.

**Comment by scottwittenburg** [comment]
I guess the source for `sz` isn't in the source mirror and currently all fetchers for that package are failing, hence the gitlab pipeline failure.

```
$ curl -OL https://github.com/szcompressor/SZ/releases/download/v2.0.2.0/sz-2.0.2.0.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100     9  100     9    0     0     17      0 --:--:-- --:--:-- --:--:--    17
$ ls -al sz-2.0.2.0.tar.gz 
-rw-r--r--  1 scott  wheel  9 Oct  6 11:11 sz-2.0.2.0.tar.gz
$ cat sz-2.0.2.0.tar.gz 
Not Found
```

**Comment by scottwittenburg** [comment]
@eugeneswalker See the comment above for the `sz` issue we were just discussing.

**Comment by opadron** [comment]
Can we get this merged?  The issue that this PR fixes is currently a blocker for work we are doing on the Viz and IO SDKs.

**Comment by tgamblin** [comment]
Is the pipeline failure here real or no?  If it's real what's the fix?

File: lib/spack/spack/binary_distribution.py [hunk 1]
```diff
@@ -350,11 +350,18 @@ def generate_key_index(key_prefix, tmpdir=None):
         for entry in web_util.list_url(key_prefix, recursive=False)
         if entry.endswith('.pub'))
 
+    remove_tmpdir = False
+    tmp = None
```
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: you shouldn't need this extra variable... just use `tmpdir`

**Review by tgamblin — CHANGES_REQUESTED** [review]

**Comment by scottwittenburg** [comment]
> Is the pipeline failure here real or no?

If you mean the gitlab pipeline failure, then no, that was a problem that came up with `sz` (#19213) which has since been fixed.

**Review by tgamblin — APPROVED** [review]

**Comment by tgamblin** [comment]
LGTM!  feel free to merge.

**Pull request merged**